src/Registration/Registration.tsx

```diff
@@ -14,76 +15,60 @@ type DispatchProps = {
 
 type Props = StateProps & DispatchProps;
 
-type State = {
-  email: string;
-  firstName: string;
-  surname: string;
-  type: UserType | null;
-};
-
-export class _Registration extends Component<Props, State> {
-  public state: State = {
-    email: '',
-    firstName: '',
-    surname: '',
-    type: null,
-  };
-
-  public render() {
-    return (
-      <div>
-        <input
-          value={this.state.firstName}
-          placeholder='firstname'
-          onChange={this.onFirstNameChange}
-        />
-        <input
-          value={this.state.surname}
-          placeholder='surname'
-          onChange={this.onSurnameChange}
-        />
-        <input
-          value={this.state.email}
-          placeholder='email'
-          onChange={this.onEmailChange}
-        />
-        <select value={this.state.type || ''} onChange={this.onTypeChange}>
-          <option value=''>Select...</option>
-          {Object.values(UserType).map(userType => (
-            <option value={userType}>{userType}</option>
-          ))}
-        </select>
-        <button onClick={this.onAddUser}>Submit</button>
-      </div>
-    );
-  }
-
-  private onTypeChange: ChangeEventHandler<HTMLSelectElement> = event => {
-    const value: UserType | null =
-      (event.currentTarget.value as UserType) || null;
-    this.setState({ type: value });
-  };
-
-  private onFirstNameChange: ChangeEventHandler<HTMLInputElement> = event => {
-    this.setState({ firstName: event.currentTarget.value });
-  };
-
-  private onSurnameChange: ChangeEventHandler<HTMLInputElement> = event => {
-    this.setState({ surname: event.currentTarget.value });
-  };
+const _Registration = (props: Props) => {
+  const [firstName, onFirstNameChange] = useField('');
+  const [surname, onSurnameChange] = useField('');
+  const [type, onTypeChange] = useField<UserType | null, HTMLSelectElement>(null);
+  const [email, onEmailChange] = useField('');
+  const [password, onPasswordChange] = useField('');
+  const [confirmPassword, onConfirmPasswordChange] = useField('');
+  const [isValid, setIsValid] = useState(false);
 
-  private onEmailChange: ChangeEventHandler<HTMLInputElement> = event => {
-    this.setState({ email: event.currentTarget.value });
-  };
-
-  private onAddUser = () => {
-    const user = { ...this.state };
+  const onAddUser = useCallback(() => {
+    const user = { email, firstName, surname, type, password, confirmPassword };
     if (user.type !== null) {
       // Typescript not registering null check -> as User
-      return this.props.addUser(user as User);
+      return props.addUser(user as User);
     }
-  };
-}
+  }, [email, firstName, surname, type, password, confirmPassword]);
+
+  const disabled =
```

Comment:
Why not use the `isValid` to determine if the button should be disabled?
The `useEffect` below needs to be a bit more complex, but then it'll align well with whether the button is enabled/disabled and reads more idiomatically.

Reply:
I've taken your suggestion and moved the `const disabled` block into my useEffect to construct a more flexible and fleshed out validity check. I agree that this only makes sense as opposed to have 2 blocks of code to execute a single process.

Reply:
This'll also work nicely when (if) you have async validation, like checking if the email already exists on the server as the user edits the form.

Reply:
I'll speak to Steve this evening about setting up a basic server to persist my `users{}` to. In the mean time I'll run this check locally but users won't survive post browser refresh.

Reply:
Why not persist it in `localStorage`, for now? That's what mostly happens during local development - don't want to rely on a server being up just so you can develop locally...

Reply:
I haven't done this as discussed